<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#项目背景" data-toc-modified-id="项目背景-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>项目背景</a></span></li><li><span><a href="#项目目标" data-toc-modified-id="项目目标-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>项目目标</a></span><ul class="toc-item"><li><span><a href="#分析目的" data-toc-modified-id="分析目的-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>分析目的</a></span></li><li><span><a href="#分析思路" data-toc-modified-id="分析思路-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>分析思路</a></span></li></ul></li><li><span><a href="#数据准备与清洗" data-toc-modified-id="数据准备与清洗-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>数据准备与清洗</a></span><ul class="toc-item"><li><span><a href="#数据字段说明" data-toc-modified-id="数据字段说明-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>数据字段说明</a></span></li><li><span><a href="#缺失值处理" data-toc-modified-id="缺失值处理-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>缺失值处理</a></span></li><li><span><a href="#数据全字段去重" data-toc-modified-id="数据全字段去重-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>数据全字段去重</a></span></li><li><span><a href="#统计每个字段的取值个数" data-toc-modified-id="统计每个字段的取值个数-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>统计每个字段的取值个数</a></span></li><li><span><a href="#时间戳处理" data-toc-modified-id="时间戳处理-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>时间戳处理</a></span></li></ul></li><li><span><a href="#数据分析" data-toc-modified-id="数据分析-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>数据分析</a></span><ul class="toc-item"><li><span><a href="#描述分析" data-toc-modified-id="描述分析-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>描述分析</a></span><ul class="toc-item"><li><span><a href="#整体维度--查看观察窗口内UV、PV、人均浏览次数、成交量的总体情况" data-toc-modified-id="整体维度--查看观察窗口内UV、PV、人均浏览次数、成交量的总体情况-4.1.1"><span class="toc-item-num">4.1.1&nbsp;&nbsp;</span>整体维度--查看观察窗口内UV、PV、人均浏览次数、成交量的总体情况</a></span></li><li><span><a href="#时间维度--按日期维度查看UV、PV、人均浏览次数、成交量" data-toc-modified-id="时间维度--按日期维度查看UV、PV、人均浏览次数、成交量-4.1.2"><span class="toc-item-num">4.1.2&nbsp;&nbsp;</span>时间维度--按日期维度查看UV、PV、人均浏览次数、成交量</a></span><ul class="toc-item"><li><span><a href="#日均UV和PV变化折线图" data-toc-modified-id="日均UV和PV变化折线图-4.1.2.1"><span class="toc-item-num">4.1.2.1&nbsp;&nbsp;</span>日均UV和PV变化折线图</a></span></li></ul></li><li><span><a href="#用户维度" data-toc-modified-id="用户维度-4.1.3"><span class="toc-item-num">4.1.3&nbsp;&nbsp;</span>用户维度</a></span><ul class="toc-item"><li><span><a href="#查看每个用户的行为数据" data-toc-modified-id="查看每个用户的行为数据-4.1.3.1"><span class="toc-item-num">4.1.3.1&nbsp;&nbsp;</span>查看每个用户的行为数据</a></span></li><li><span><a href="#查看购买用户数、复购用户数以及复购率" data-toc-modified-id="查看购买用户数、复购用户数以及复购率-4.1.3.2"><span class="toc-item-num">4.1.3.2&nbsp;&nbsp;</span>查看购买用户数、复购用户数以及复购率</a></span></li><li><span><a href="#用户跳失率" data-toc-modified-id="用户跳失率-4.1.3.3"><span class="toc-item-num">4.1.3.3&nbsp;&nbsp;</span>用户跳失率</a></span></li></ul></li></ul></li><li><span><a href="#用户流量行为转化分析" data-toc-modified-id="用户流量行为转化分析-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>用户流量行为转化分析</a></span></li><li><span><a href="#用户行为习惯分析" data-toc-modified-id="用户行为习惯分析-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>用户行为习惯分析</a></span></li><li><span><a href="#用户类目偏好分析" data-toc-modified-id="用户类目偏好分析-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>用户类目偏好分析</a></span><ul class="toc-item"><li><span><a href="#查看pv—top20的商品类目" data-toc-modified-id="查看pv—top20的商品类目-4.4.1"><span class="toc-item-num">4.4.1&nbsp;&nbsp;</span>查看pv—top20的商品类目</a></span></li><li><span><a href="#将上诉过程定义为函数" data-toc-modified-id="将上诉过程定义为函数-4.4.2"><span class="toc-item-num">4.4.2&nbsp;&nbsp;</span>将上诉过程定义为函数</a></span></li><li><span><a href="#比较所有行为的top20的商品类别" data-toc-modified-id="比较所有行为的top20的商品类别-4.4.3"><span class="toc-item-num">4.4.3&nbsp;&nbsp;</span>比较所有行为的top20的商品类别</a></span></li></ul></li></ul></li><li><span><a href="#结论" data-toc-modified-id="结论-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>结论</a></span></li></ul></div>

# 项目背景  
UserBehavior是阿里巴巴提供的一个淘宝用户行为数据集，数据集包含了2017年11月25日至2017年12月3日之间，有行为的约一百万随机用户的所有行为（行为包括点击、购买、加购、喜欢），用以进行隐式反馈推荐问题的研究。  
注：隐式反馈推荐问题
通过对2017年11月25日至2017年12月3日之间的用户行为数据分析，为客户提供更精准的隐式反馈推荐。  

# 项目目标

## 分析目的

**1.从用户角度：提高用户忠诚度，帮助用户快速找到商品，减轻信息过载带来的问题。**   
**2.从网站角度：提高网站交叉销售能力，提高成交转化率， 从而提高成交额。**  
**3.最终目的：能使得推荐系统能将更多的物品推荐给用户， 能给用户推荐他们真正感兴趣的物品。**  
注：数据来源为下面所示的天池数据集，但是由于数据量过大（约有1亿条用户行为），所以抽样约**10000000**的用户行为进行分析。 

参考资料  
> https://zhuanlan.zhihu.com/p/105766780 《数据分析项目实战——淘宝用户行为分析》-------作者：（知乎）何书365

## 分析思路  
主要进行以下四个专题分析：
1. <font color=red>描述分析</font>:从整体维度和日期维度查看观察窗口内的UV PV PV/UV 成交量， 从用户维度查看用户的行为数据。
2. <font color=red>用户流量行为分析</font>:分析网站用户的各行为转化率。  
3. <font color=red>用户类目偏好分析</font>:从商品的点击、收藏、加购、购买的频率，探索用户对商品的购买偏好，找到针对不同商品的营销策略(什么产品以及产品类目的购买率最高，找出最受欢迎的产品，优化产品推荐)。   
4. <font color=red>用户行为习惯分析</font>:从时间维度分析观察窗口内平均而言一天当中的哪段时间用户数最多， 找到一天当中用户最活跃的时刻。


# 数据准备与清洗

In [1]:
#全部行都能输出
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

In [284]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
plt.style.use('seaborn')
#%matplotlib inline  
import warnings
warnings.filterwarnings("ignore")

In [285]:
pd.set_option('max_rows', 100) 
pd.set_option('max_columns', 50) 

In [286]:
plt.rcParams['font.sans-serif']=['Simhei']  #显示中文,解决图中无法显示中文的问题
plt.rcParams['axes.unicode_minus']=False    #设置显示中文后,负号显示受影响。解决坐标轴上乱码问题 

In [5]:
# csv的压缩数据
# header=None不以第一行作为标题
df = pd.read_csv(r"D:\资料\项目实战\淘宝用户行为数据分析\data\UserBehavior.csv.zip", compression='zip', iterator=True, header=None)
df
# 由于这里数据量很大， 直接全部加载进来会直接占用内存， 故这里只抽取20000000行数据做分析
#  iterator=True是产生一个TextFileReader对象， 不会直接全部把数据读进来
# 如果想用所有数据做分析， 就利用chunksize=int, 指定生成一个迭代器， 每次指定读取int行数据， 批量进行处理也行

In [6]:
# TextFileReader对象的get_chunk(10000000)指定获取10000000行数据
df = df.get_chunk(10000000)

In [7]:
df.shape  # 一共10000000， 5列

(10000000, 5)

In [8]:
df.head()

0        1        2   3           4
0  1  2268318  2520377  pv  1511544070
1  1  2333346  2520771  pv  1511561733
2  1  2576651   149192  pv  1511572885
3  1  3830808  4181361  pv  1511593493
4  1  4365585  2520377  pv  1511596146

In [9]:
# 查看数据信息
df.info()

# 抽取出来的1000万行数据有381.5+ MB， 其实不是很大， 虽然行数很多， 但是列数只有5列， 数据类型都很简单， 没有长文本
# 所以完全可以考虑分析更大量的数据， 此处， 我们就以1000万行数据分析即可

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000000 entries, 0 to 9999999
Data columns (total 5 columns):
 #   Column  Dtype 
---  ------  ----- 
 0   0       int64 
 1   1       int64 
 2   2       int64 
 3   3       object
 4   4       int64 
dtypes: int64(4), object(1)
memory usage: 381.5+ MB


## 数据字段说明  
本数据集包含了2017年11月25日至2017年12月3日之间，有行为的**约一百万**随机用户的所有行为（行为包括点击、购买、加购、喜欢）。数据集的组织形式和MovieLens-20M类似，即数据集的每一行表示一条用户行为，由用户ID、商品ID、商品类目ID、行为类型和时间戳组成，并以逗号分隔。关于数据集中每一列的详细描述如下：

|    | 列名称     | 说明                                               |
|---:|:-----------|:---------------------------------------------------|
|  0 | 用户ID     | 整数类型，序列化后的用户ID                         |
|  1 | 商品ID     | 整数类型，序列化后的商品ID                         |
|  2 | 商品类目ID | 整数类型，序列化后的商品所属类目ID                 |
|  3 | 行为类型   | 字符串，枚举类型，包括('pv', 'buy', 'cart', 'fav') |
|  4 | 时间戳     | 行为发生的时间戳                                   |

In [10]:
# 添加列名
columns = ['用户ID', '商品ID', '商品类目ID', '行为类型', '时间戳']
columns

['用户ID', '商品ID', '商品类目ID', '行为类型', '时间戳']

In [11]:
df.columns = columns

In [12]:
df

用户ID     商品ID   商品类目ID 行为类型         时间戳
0             1  2268318  2520377   pv  1511544070
1             1  2333346  2520771   pv  1511561733
2             1  2576651   149192   pv  1511572885
3             1  3830808  4181361   pv  1511593493
4             1  4365585  2520377   pv  1511596146
...         ...      ...      ...  ...         ...
9999995  536993    95420  3253210   pv  1511934240
9999996  536993  1702210  1689352   pv  1511934277
9999997  536993  3023810  2832009   pv  1511934294
9999998  536993  1702210  1689352   pv  1511934330
9999999  536993  4505862  1689352   pv  1511934431

[10000000 rows x 5 columns]

## 缺失值处理

In [13]:
df[df.isna().values]  # 无缺失值

Empty DataFrame
Columns: [用户ID, 商品ID, 商品类目ID, 行为类型, 时间戳]
Index: []

## 数据全字段去重

In [14]:
# 提取出重复的行
df[df.duplicated()]  # 有5行是重复的行

用户ID     商品ID   商品类目ID 行为类型         时间戳
2293745   18711  3324300  1464116  buy  1511676439
3542227  242650  3967698   796856  buy  1512097481
4294120  277134  3404251   149192  buy  1512224835
4294121  277134  1697974   982926  buy  1512224835
4849518   30295  1081025   895939  buy  1511594724

In [15]:
# 删除重复的行
df.drop_duplicates(inplace=True)

In [16]:
df.shape

(9999995, 5)

In [17]:
# 重置行索引
df.reset_index(drop=True, inplace=True)

In [18]:
df.head()

用户ID     商品ID   商品类目ID 行为类型         时间戳
0     1  2268318  2520377   pv  1511544070
1     1  2333346  2520771   pv  1511561733
2     1  2576651   149192   pv  1511572885
3     1  3830808  4181361   pv  1511593493
4     1  4365585  2520377   pv  1511596146

In [19]:
df.shape

(9999995, 5)

## 统计每个字段的取值个数

In [20]:
df.apply(lambda x:x.value_counts().size)

用户ID        97811
商品ID      1561016
商品类目ID       7967
行为类型            4
时间戳        760501
dtype: int64

从中可以看出：
1. 所用数据中的用户一共有97811个
2. 商品共有1561016个
3. 商品一共有7967类
4. 行为类型有四个

In [21]:
# 对行为类型进行统计
df.行为类型.value_counts()

pv      8950064
cart     559132
fav      291657
buy      199142
Name: 行为类型, dtype: int64

注意到，**用户行为类型 共有 四种**，它们分别是:   

|    | 行为类型   |    计数 | 说明                     |
|---:|:-----------|--------:|:-------------------------|
|  0 | pv         | 8950064 | 商品详情页pv，等价于点击 |
|  1 | buy        |  199142 | 商品购买                 |
|  2 | cart       |  559132 | 将商品加入购物车         |
|  3 | fav        |  291657 | 收藏商品                 |

In [22]:
df.head()

用户ID     商品ID   商品类目ID 行为类型         时间戳
0     1  2268318  2520377   pv  1511544070
1     1  2333346  2520771   pv  1511561733
2     1  2576651   149192   pv  1511572885
3     1  3830808  4181361   pv  1511593493
4     1  4365585  2520377   pv  1511596146

## 时间戳处理  
在计算机中，时间实际上是用数字表示的。  
我们把1970年1月1日 00:00:00 UTC+00:00时区的时刻称为epoch time(纪元)，记为0（1970年以前的时间timestamp为负数），当前时间就是相对于epoch time的秒数，称为timestamp。   
timestamp的值与时区毫无关系，因为timestamp一旦确定，其UTC时间就确定了，转换到任意时区的时间也是完全确定的，这就是为什么计算机存储的当前时间是以timestamp表示的，因为全球各地的计算机在任意时刻的timestamp都是完全相同的（假定时间已校准）

In [23]:
import time
time.time()  # 现在的时间戳， # 整数位10是秒， 即1970年1月1日 00:00:00以来的秒数

1597544038.7362134

In [24]:
# 把时间戳转换成日期时间的方法
pd.Timestamp(time.time(), unit='s')   # 日期是对的， 但是时区不对， 需要加上8个小时

Timestamp('2020-08-16 02:14:00.013303280')

In [25]:
# 将时间戳转换成日期时间
pd.to_datetime(df.时间戳, unit='s')

0         2017-11-24 17:21:10
1         2017-11-24 22:15:33
2         2017-11-25 01:21:25
3         2017-11-25 07:04:53
4         2017-11-25 07:49:06
                  ...        
9999990   2017-11-29 05:44:00
9999991   2017-11-29 05:44:37
9999992   2017-11-29 05:44:54
9999993   2017-11-29 05:45:30
9999994   2017-11-29 05:47:11
Name: 时间戳, Length: 9999995, dtype: datetime64[ns]

In [26]:
# 分离日期和时间
a = pd.to_datetime(df.时间戳, unit='s').astype(str).str.split(' ', expand=True)
a

0         1
0        2017-11-24  17:21:10
1        2017-11-24  22:15:33
2        2017-11-25  01:21:25
3        2017-11-25  07:04:53
4        2017-11-25  07:49:06
...             ...       ...
9999990  2017-11-29  05:44:00
9999991  2017-11-29  05:44:37
9999992  2017-11-29  05:44:54
9999993  2017-11-29  05:45:30
9999994  2017-11-29  05:47:11

[9999995 rows x 2 columns]

In [27]:
a.columns = ['日期', '时间']

In [28]:
a

日期        时间
0        2017-11-24  17:21:10
1        2017-11-24  22:15:33
2        2017-11-25  01:21:25
3        2017-11-25  07:04:53
4        2017-11-25  07:49:06
...             ...       ...
9999990  2017-11-29  05:44:00
9999991  2017-11-29  05:44:37
9999992  2017-11-29  05:44:54
9999993  2017-11-29  05:45:30
9999994  2017-11-29  05:47:11

[9999995 rows x 2 columns]

In [29]:
# 拼接到原表中
df = pd.concat([df, a], axis=1)

In [30]:
df.head()

用户ID     商品ID   商品类目ID 行为类型         时间戳          日期        时间
0     1  2268318  2520377   pv  1511544070  2017-11-24  17:21:10
1     1  2333346  2520771   pv  1511561733  2017-11-24  22:15:33
2     1  2576651   149192   pv  1511572885  2017-11-25  01:21:25
3     1  3830808  4181361   pv  1511593493  2017-11-25  07:04:53
4     1  4365585  2520377   pv  1511596146  2017-11-25  07:49:06

In [31]:
# 对日期进行分类计数
df.日期.value_counts()   # 我们只取2017年11月25日至2017年12月3日之间的数据

2017-12-02    1399859
2017-12-03    1201017
2017-12-01    1122808
2017-11-26    1053119
2017-11-30    1046205
2017-11-25    1045084
2017-11-29    1024392
2017-11-27     999106
2017-11-28     984832
2017-11-24     122028
2017-11-23        578
2017-11-22        183
2017-11-21         84
2017-11-03         70
2017-11-19         66
2017-11-20         62
2017-11-18         56
2017-11-17         37
1905-07-13         28
2019-02-18         24
2017-11-12         23
2017-11-11         19
2017-04-13         18
2016-10-20         17
2018-08-28         16
2017-11-16         16
2020-05-09         15
2019-02-25         14
2019-02-24         12
2017-10-26         12
2012-12-03         12
2017-04-12         12
2025-10-11         11
2017-11-14         11
2020-10-18         11
2017-12-06         11
2019-02-20         11
2017-04-11          9
2019-02-26          9
2017-11-15          9
2017-10-29          8
2017-12-04          8
2018-06-20          6
2019-02-23          6
2017-11-10          6
2017-11-13

In [32]:
a = pd.to_datetime(df.日期)
a

0         2017-11-24
1         2017-11-24
2         2017-11-25
3         2017-11-25
4         2017-11-25
             ...    
9999990   2017-11-29
9999991   2017-11-29
9999992   2017-11-29
9999993   2017-11-29
9999994   2017-11-29
Name: 日期, Length: 9999995, dtype: datetime64[ns]

In [33]:
a = a.apply(lambda x:x>=pd.Timestamp('2017-11-25') and x<=pd.Timestamp('2017-12-03'))
a

0          False
1          False
2           True
3           True
4           True
           ...  
9999990     True
9999991     True
9999992     True
9999993     True
9999994     True
Name: 日期, Length: 9999995, dtype: bool

In [36]:
df = df[a]  # 取满足条件的行

In [37]:
df

用户ID     商品ID   商品类目ID 行为类型         时间戳          日期        时间
2             1  2576651   149192   pv  1511572885  2017-11-25  01:21:25
3             1  3830808  4181361   pv  1511593493  2017-11-25  07:04:53
4             1  4365585  2520377   pv  1511596146  2017-11-25  07:49:06
5             1  4606018  2735466   pv  1511616481  2017-11-25  13:28:01
6             1   230380   411153   pv  1511644942  2017-11-25  21:22:22
...         ...      ...      ...  ...         ...         ...       ...
9999990  536993    95420  3253210   pv  1511934240  2017-11-29  05:44:00
9999991  536993  1702210  1689352   pv  1511934277  2017-11-29  05:44:37
9999992  536993  3023810  2832009   pv  1511934294  2017-11-29  05:44:54
9999993  536993  1702210  1689352   pv  1511934330  2017-11-29  05:45:30
9999994  536993  4505862  1689352   pv  1511934431  2017-11-29  05:47:11

[9876422 rows x 7 columns]

In [38]:
# 再对日期进行统计检查一下
df.日期.value_counts()

2017-12-02    1399859
2017-12-03    1201017
2017-12-01    1122808
2017-11-26    1053119
2017-11-30    1046205
2017-11-25    1045084
2017-11-29    1024392
2017-11-27     999106
2017-11-28     984832
Name: 日期, dtype: int64

In [39]:
# 重置行索引
df.reset_index(drop=True, inplace=True)

In [40]:
df  # 现在只剩下9876422行

用户ID     商品ID   商品类目ID 行为类型         时间戳          日期        时间
0             1  2576651   149192   pv  1511572885  2017-11-25  01:21:25
1             1  3830808  4181361   pv  1511593493  2017-11-25  07:04:53
2             1  4365585  2520377   pv  1511596146  2017-11-25  07:49:06
3             1  4606018  2735466   pv  1511616481  2017-11-25  13:28:01
4             1   230380   411153   pv  1511644942  2017-11-25  21:22:22
...         ...      ...      ...  ...         ...         ...       ...
9876417  536993    95420  3253210   pv  1511934240  2017-11-29  05:44:00
9876418  536993  1702210  1689352   pv  1511934277  2017-11-29  05:44:37
9876419  536993  3023810  2832009   pv  1511934294  2017-11-29  05:44:54
9876420  536993  1702210  1689352   pv  1511934330  2017-11-29  05:45:30
9876421  536993  4505862  1689352   pv  1511934431  2017-11-29  05:47:11

[9876422 rows x 7 columns]

In [41]:
df.to_csv('清洗好的数据.csv',index=False)

# 数据分析

## 描述分析

### 整体维度--查看观察窗口内UV、PV、人均浏览次数、成交量的总体情况

In [45]:
df

用户ID     商品ID   商品类目ID 行为类型         时间戳          日期        时间
0             1  2576651   149192   pv  1511572885  2017-11-25  01:21:25
1             1  3830808  4181361   pv  1511593493  2017-11-25  07:04:53
2             1  4365585  2520377   pv  1511596146  2017-11-25  07:49:06
3             1  4606018  2735466   pv  1511616481  2017-11-25  13:28:01
4             1   230380   411153   pv  1511644942  2017-11-25  21:22:22
...         ...      ...      ...  ...         ...         ...       ...
9876417  536993    95420  3253210   pv  1511934240  2017-11-29  05:44:00
9876418  536993  1702210  1689352   pv  1511934277  2017-11-29  05:44:37
9876419  536993  3023810  2832009   pv  1511934294  2017-11-29  05:44:54
9876420  536993  1702210  1689352   pv  1511934330  2017-11-29  05:45:30
9876421  536993  4505862  1689352   pv  1511934431  2017-11-29  05:47:11

[9876422 rows x 7 columns]

In [46]:
UV = len(df.用户ID.unique())
PV = df.行为类型.value_counts()['pv']
pd.DataFrame(pd.Series({'UV' :UV, 
                        'PV' :PV, 
                        '人均浏览次数' : PV/UV, 
                        '成交量' : df.行为类型.value_counts()['buy'] 
             
                        })).T

UV         PV     人均浏览次数       成交量
0  97809.0  8838199.0  90.361817  197534.0

### 时间维度--按日期维度查看UV、PV、人均浏览次数、成交量

1. UV = Unique Visitor独立访客, 每个日期下的不同用户ID计数
2. PV = page view页面浏览量, 每个日期下pv的计数
3. 人均浏览次数 = PV/UV
4. 成交量 = 每个日期buy的求和

In [47]:
def data_UV_PV_UVPV_cj(DF):
    # 独立访客
    UV = DF['用户ID'].value_counts().size
    # 页面浏览量
    PV = DF['行为类型'].value_counts()['pv']
    # 人均浏览次数
    人均浏览次数 = float(format(PV/UV, '.2f'))
    # 成交量
    成交量 = DF['行为类型'].value_counts()['buy']
    return pd.Series({'UV':UV, 
                         'PV':PV, 
                         '人均浏览次数':人均浏览次数, 
                         '成交量':成交量
                        })
a = df.groupby('日期').apply(data_UV_PV_UVPV_cj)

In [48]:
a

UV         PV  人均浏览次数      成交量
日期                                             
2017-11-25  69730.0   938025.0   13.45  19694.0
2017-11-26  70690.0   942188.0   13.33  21060.0
2017-11-27  70170.0   893529.0   12.73  21408.0
2017-11-28  70177.0   881044.0   12.55  20953.0
2017-11-29  71210.0   916202.0   12.87  21834.0
2017-11-30  72459.0   935968.0   12.92  21836.0
2017-12-01  74572.0  1004899.0   13.48  21285.0
2017-12-02  93371.0  1251569.0   13.40  26049.0
2017-12-03  90887.0  1074775.0   11.83  23415.0

In [49]:
# 导出数据
a.reset_index().to_csv('UV_PV_GBK.csv', index=False, encoding='gbk')

#### 日均UV和PV变化折线图

In [93]:
# 利用pyecharts进行可视化, 用matplotlib更简单， 但是图不怎么好看
# 导入全局配置项
from pyecharts import options as opts
# 导入背景主题类
from pyecharts.globals import ThemeType
# 导入线图的类
from pyecharts.charts import Line

In [315]:
# 实例化对象
line = Line(init_opts=opts.InitOpts(theme=ThemeType.MACARONS, width='1000px', height='330px'))
# x轴
line.add_xaxis(a.index.to_list())
# y轴
line.add_yaxis(series_name='UV', 
               y_axis=a.UV.values.tolist(), 
               color='green', 
               symbol='pin', # 标记
               symbol_size=20, #标记的大小
               # 标记线配置项
               markline_opts = opts.MarkLineOpts(data = [opts.MarkLineItem(type_ ="average" , name = '平均值')]), 
               # 标记点配置项
               markpoint_opts = opts.MarkPointOpts(data = [opts.MarkPointItem(type_='max', name = '最大值', 
                                                           symbol = 'circle'
                                                                             )])
              )
line.add_yaxis(series_name='PV', 
               y_axis=a.PV.values.tolist(), 
               color='green', 
               symbol='diamond', # 标记
               symbol_size=20, #标记的大小
               # 标记线配置项
               markline_opts = opts.MarkLineOpts(data = [opts.MarkLineItem(type_ ="average" , name = '平均值')]), 
               # 标记点配置项
               markpoint_opts = opts.MarkPointOpts(data = [opts.MarkPointItem(type_='max', name = '最大值', 
                                                           symbol = 'diamond'
                                                                             )])
              )
# 全局配置项          # 标题
line.set_global_opts( title_opts = opts.TitleOpts(title = '日均UV和PV变化趋势线', pos_left=30), 
                      # x轴配置                  # x轴名字
                      xaxis_opts = opts.AxisOpts(name='日期', 
                                                 # 名字所放的位置
                                                 name_location='end', 
                                                 # 坐标轴名称与轴线之间的距离
                                                 name_gap=20
                                                ), 
                     # y轴配置
                     yaxis_opts = opts.AxisOpts(name='数量')
                    
                    )
line.render_notebook()

D:\Anaconda3\lib\site-packages\pyecharts\charts\chart.py:14: PendingDeprecationWarning: pyecharts 所有图表类型将在 v1.9.0 版本开始强制使用 ChartItem 进行数据项配置 :)
  super().__init__(init_opts=init_opts)


2017年12月1号到2号， 数据明显向上波动，查了查可能是因为双12领红包优惠券开始时间是12月1号开始   
所以导致这天pv上升明显

### 用户维度

#### 查看每个用户的行为数据

In [50]:
df.head(3)

用户ID     商品ID   商品类目ID 行为类型         时间戳          日期        时间
0     1  2576651   149192   pv  1511572885  2017-11-25  01:21:25
1     1  3830808  4181361   pv  1511593493  2017-11-25  07:04:53
2     1  4365585  2520377   pv  1511596146  2017-11-25  07:49:06

In [8]:
a = df.groupby('用户ID').apply(lambda x:x.行为类型.value_counts())

In [11]:
a = a.unstack()

In [13]:
a.columns = ['购买', '加入购物车', '收藏', '点击浏览']   # 出现很多缺失值是正常的， 比如用户1， 只看， 其他什么也没做
    # 再比如用户13， 处了， 看， 还加入购物车了。

In [17]:
a.sort_index(inplace=True)

In [18]:
a

购买  加入购物车   收藏   点击浏览
用户ID                           
1        NaN    NaN  NaN   53.0
5        NaN    NaN  NaN   64.0
13       NaN    4.0  NaN    7.0
19       NaN    1.0  NaN   74.0
21       NaN    1.0  6.0  310.0
...      ...    ...  ...    ...
1017990  NaN    3.0  NaN   56.0
1017994  NaN    3.0  3.0   29.0
1017997  2.0    NaN  3.0   80.0
1018000  NaN    4.0  NaN  177.0
1018011  1.0    NaN  NaN   41.0

[97809 rows x 4 columns]

In [20]:
df.用户ID.value_counts().sort_index()  # 用户行为数， 并按照用户ID进行升序

1           53
5           64
13          11
19          75
21         317
          ... 
1017990     59
1017994     35
1017997     85
1018000    181
1018011     42
Name: 用户ID, Length: 97809, dtype: int64

In [21]:
a['用户行为数'] = df.用户ID.value_counts().sort_index()

In [22]:
a  # 缺失的地方代表用户没有该行为， 可以用0代替缺失值

购买  加入购物车   收藏   点击浏览  用户行为数
用户ID                                  
1        NaN    NaN  NaN   53.0     53
5        NaN    NaN  NaN   64.0     64
13       NaN    4.0  NaN    7.0     11
19       NaN    1.0  NaN   74.0     75
21       NaN    1.0  6.0  310.0    317
...      ...    ...  ...    ...    ...
1017990  NaN    3.0  NaN   56.0     59
1017994  NaN    3.0  3.0   29.0     35
1017997  2.0    NaN  3.0   80.0     85
1018000  NaN    4.0  NaN  177.0    181
1018011  1.0    NaN  NaN   41.0     42

[97809 rows x 5 columns]

In [24]:
# 导出数据 a.reset_index().to_csv('用户行为数据.csv', index=False)

#### 查看购买用户数、复购用户数以及复购率  
复购的定义：这里复购是对网站而不是商品而言， 只要有2次及以上的购买行为， 我们可以称该用户为复购客户。  
复购率的定义：是指有两次及以上购买行为的用户数/有购买行为的用户数

In [52]:
# 节省内存， 重新启动下， 所以这里， 我们重新导入数据
a = pd.read_csv('用户行为数据.csv')
a

用户ID   购买  加入购物车   收藏   点击浏览  用户行为数
0            1  NaN    NaN  NaN   53.0     53
1            5  NaN    NaN  NaN   64.0     64
2           13  NaN    4.0  NaN    7.0     11
3           19  NaN    1.0  NaN   74.0     75
4           21  NaN    1.0  6.0  310.0    317
...        ...  ...    ...  ...    ...    ...
97804  1017990  NaN    3.0  NaN   56.0     59
97805  1017994  NaN    3.0  3.0   29.0     35
97806  1017997  2.0    NaN  3.0   80.0     85
97807  1018000  NaN    4.0  NaN  177.0    181
97808  1018011  1.0    NaN  NaN   41.0     42

[97809 rows x 6 columns]

In [53]:
# 填充缺失值
a.fillna(0, inplace=True)

In [55]:
a

用户ID   购买  加入购物车   收藏   点击浏览  用户行为数
0            1  0.0    0.0  0.0   53.0     53
1            5  0.0    0.0  0.0   64.0     64
2           13  0.0    4.0  0.0    7.0     11
3           19  0.0    1.0  0.0   74.0     75
4           21  0.0    1.0  6.0  310.0    317
...        ...  ...    ...  ...    ...    ...
97804  1017990  0.0    3.0  0.0   56.0     59
97805  1017994  0.0    3.0  3.0   29.0     35
97806  1017997  2.0    0.0  3.0   80.0     85
97807  1018000  0.0    4.0  0.0  177.0    181
97808  1018011  1.0    0.0  0.0   41.0     42

[97809 rows x 6 columns]

In [57]:
pd.DataFrame(pd.Series(dict(购买用户数 = (a.购买>0).sum(),
                            复购用户数 = (a.购买>1).sum(),
                            复购率 = (a.购买>1).sum()/(a.购买>0).sum()))).T

购买用户数    复购用户数       复购率
0  66248.0  43632.0  0.658616

#### 用户跳失率  
跳失率的定义， 用户更具链接点击进入淘宝网页， 我们并不知道统计规则里面这个进来的页面是否会算作pv  
安全起见，我们认为这个算作pv， 但不是主动浏览的， 可能是广告渠道等等。  
所以， 我们定义， **只有pv>=2的客户才不是跳失客户**。但是很奇怪的是有些用户pv没有， 但是却有其他行为， 甚至是购买行为， 这些客户就不能算是跳失客户， 所以， 还同时要求其他行为均为0.

In [58]:
a.head()

用户ID   购买  加入购物车   收藏   点击浏览  用户行为数
0     1  0.0    0.0  0.0   53.0     53
1     5  0.0    0.0  0.0   64.0     64
2    13  0.0    4.0  0.0    7.0     11
3    19  0.0    1.0  0.0   74.0     75
4    21  0.0    1.0  6.0  310.0    317

In [59]:
a[(a.点击浏览<2)&(a.购买==0)&(a.加入购物车==0)&(a.收藏==0)]

# 也就是说有3个客户刚点进来就退出网页了

用户ID   购买  加入购物车   收藏  点击浏览  用户行为数
74891  436370  0.0    0.0  0.0   1.0      1
79195  458889  0.0    0.0  0.0   1.0      1
85372  490686  0.0    0.0  0.0   1.0      1

In [60]:
pd.DataFrame(pd.Series(dict(跳失率 = format(3/len(a), '.5%'))))

0
跳失率  0.00307%

当统计时间为上诉观察窗口时，用户的复购率为65.86%且跳失率接近0，说明淘宝拥有你足够的吸引力让用户停留且用户的忠诚度非常高。  
可以进一步培养用户的忠诚度，鼓励用户更高频次的消费。

## 用户流量行为转化分析

In [61]:
总浏览点击 = (a[a.点击浏览>2]['点击浏览']).sum()  # 网站一共获取的点击量
总浏览点击

8837973.0

In [62]:
总收藏 = a.收藏.sum()                       # 获取的收藏量
总收藏

287873.0

In [63]:
总加入购物车 = a.加入购物车.sum()        # 获取的加入购物车的量
总加入购物车

552816.0

In [64]:
总购买 = a.购买.sum()
总购买

197534.0

In [65]:
(总收藏+总加入购物车)/总浏览点击     # 产生购买倾向

0.09512237704279024

In [66]:
总购买/总浏览点击      

0.022350600075379274

In [76]:
import pyecharts.options as opts
from pyecharts.charts import Funnel
x_data = ["pv", "加购物车+收藏", "购买"]
y_data = [100, 95.12, 2.24]

data = [[x_data[i], y_data[i]] for i in range(len(x_data))]

c = (
    Funnel(init_opts=opts.InitOpts(width="600px", height="300px"))
    .add(
        series_name="",
        data_pair=data,
        gap=2,
        tooltip_opts=opts.TooltipOpts(trigger="item", formatter="{a} <br/>{b} : {c}%"),
        label_opts=opts.LabelOpts(is_show=True, position="inside"),
        itemstyle_opts=opts.ItemStyleOpts(border_color="#fff", border_width=1),
    )
    .set_global_opts(title_opts=opts.TitleOpts(title="流量行为漏斗图", subtitle=""))
)
c.render_notebook()

D:\Anaconda3\lib\site-packages\pyecharts\charts\chart.py:14: PendingDeprecationWarning: pyecharts 所有图表类型将在 v1.9.0 版本开始强制使用 ChartItem 进行数据项配置 :)
  super().__init__(init_opts=init_opts)


从整个流量上来说， 客户的购买倾向还是很高的， 说明商品展览等相关说明做的不错， 但是最终购买率2.24%很低，可能是要对不同店铺的同种产品进行比较

## 用户行为习惯分析  
这里主要分析一天当中那个时间段用户最活跃

In [86]:
#df = pd.read_csv('清洗好的数据.csv')
df.时间.str.split(':').str[0]

0          01
1          07
2          07
3          13
4          21
           ..
9876417    05
9876418    05
9876419    05
9876420    05
9876421    05
Name: 时间, Length: 9876422, dtype: object

In [88]:
df['时间段'] = df.时间.str.split(':').str[0]  # 警告不用管

<ipython-input-88-83f0792f57ff>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['时间段'] = df.时间.str.split(':').str[0]


In [261]:
df.head()

用户ID     商品ID   商品类目ID 行为类型         时间戳          日期        时间 时间段
0     1  2576651   149192   pv  1511572885  2017-11-25  01:21:25  01
1     1  3830808  4181361   pv  1511593493  2017-11-25  07:04:53  07
2     1  4365585  2520377   pv  1511596146  2017-11-25  07:49:06  07
3     1  4606018  2735466   pv  1511616481  2017-11-25  13:28:01  13
4     1   230380   411153   pv  1511644942  2017-11-25  21:22:22  21

In [262]:
a = df.groupby('时间段')
a

In [94]:
a = a.apply(lambda x:len(set(x.用户ID)))  # 每个时间段的UV
a

时间段
00    44182
01    51714
02    57150
03    58234
04    58843
05    59914
06    58765
07    59941
08    59385
09    57779
10    57727
11    61928
12    65920
13    67647
14    64359
15    50735
16    29232
17    14931
18     8801
19     6242
20     5647
21     7654
22    16593
23    30450
dtype: int64

In [113]:
# 实例化对象
line = Line(init_opts=opts.InitOpts(theme=ThemeType.ESSOS, width='1000px', height='330px'))
# x轴
line.add_xaxis(a.index.to_list())
# y轴
line.add_yaxis(series_name='UV', 
               y_axis=a.values.tolist(), 
               color='green', 
               symbol='pin', # 标记
               symbol_size=20, #标记的大小
               # 标记线配置项
               markline_opts = opts.MarkLineOpts(data = [opts.MarkLineItem(type_ ="average" , name = '平均值')]), 
               # 标记点配置项
               markpoint_opts = opts.MarkPointOpts(data = [opts.MarkPointItem(type_='max', name = '最大值', 
                                                           symbol = 'circle'
                                                                             )])
              )

# 全局配置项          # 标题
line.set_global_opts( title_opts = opts.TitleOpts(title = '用户行为习惯分布', pos_left=48), 
                      # x轴配置                  # x轴名字
                      xaxis_opts = opts.AxisOpts(name='时间段', 
                                                 # 名字所放的位置
                                                 name_location='end', 
                                                 # 坐标轴名称与轴线之间的距离
                                                 name_gap=20
                                                ), 
                     # y轴配置
                     yaxis_opts = opts.AxisOpts(name='数量')
                    
                    )

In [112]:
line.render_notebook()

由此可以看出， 一天当中用户数量最活跃的时候是在13:00左右，之后开始急剧降低，  最低是在晚上19:00-20:00，  
之后开始上升， 也就是说人们睡觉之前会喜欢浏览网页。所以如果要做什么活动，打什么广告最好再13:00左右进行

## 用户类目偏好分析

### 查看pv—top20的商品类目

In [241]:
a = df[df.行为类型=='pv']['商品类目ID'].value_counts()[0:20]
a

4756105    478455
4145813    322614
2355072    312967
3607361    293164
982926     283172
2520377    197556
4801426    186675
1320293    174916
2465336    154782
3002561    142784
2735466    114214
4181361    101568
149192      98842
1080785     97608
2885642     94869
4217906     93132
154040      88305
3738615     84852
1879194     77468
2640118     71869
Name: 商品类目ID, dtype: int64

In [242]:
b = a.to_frame().reset_index()
b['name'] = b.pop('index')
b = b.rename(columns = {"商品类目ID":'value'})
b = b.T.to_dict()
b = b.values()
b = list(b)
b

[{'value': 478455, 'name': 4756105},
 {'value': 322614, 'name': 4145813},
 {'value': 312967, 'name': 2355072},
 {'value': 293164, 'name': 3607361},
 {'value': 283172, 'name': 982926},
 {'value': 197556, 'name': 2520377},
 {'value': 186675, 'name': 4801426},
 {'value': 174916, 'name': 1320293},
 {'value': 154782, 'name': 2465336},
 {'value': 142784, 'name': 3002561},
 {'value': 114214, 'name': 2735466},
 {'value': 101568, 'name': 4181361},
 {'value': 98842, 'name': 149192},
 {'value': 97608, 'name': 1080785},
 {'value': 94869, 'name': 2885642},
 {'value': 93132, 'name': 4217906},
 {'value': 88305, 'name': 154040},
 {'value': 84852, 'name': 3738615},
 {'value': 77468, 'name': 1879194},
 {'value': 71869, 'name': 2640118}]

In [257]:
from pyecharts import options as opts
from pyecharts.charts import TreeMap
from pyecharts.globals import ThemeType
data =b
c = (
    TreeMap(init_opts=opts.InitOpts(theme=ThemeType.LIGHT, width='800px', height='400px'))
    .add("pv", data, )
    .set_global_opts(title_opts=opts.TitleOpts(title="pv-top20商品类目"))
)

D:\Anaconda3\lib\site-packages\pyecharts\charts\chart.py:14: PendingDeprecationWarning: pyecharts 所有图表类型将在 v1.9.0 版本开始强制使用 ChartItem 进行数据项配置 :)
  super().__init__(init_opts=init_opts)


In [258]:
c.render_notebook()

我们可以看到pv数目最高的商品类目是4756105， 为478455次, 我们再看看其他用户行为的商品类目top20

### 将上诉过程定义为函数

In [269]:
import warnings
warnings.filterwarnings("ignore")

In [270]:
def tree_plot(be):
    a = df[df.行为类型==be]['商品类目ID'].value_counts()[0:20]
    #print(a)
    b = a.to_frame().reset_index()
    b['name'] = b.pop('index')
    b = b.rename(columns = {"商品类目ID":'value'})
    b = b.T.to_dict()
    b = b.values()
    b = list(b)
    # b
    from pyecharts import options as opts
    from pyecharts.charts import TreeMap
    from pyecharts.globals import ThemeType
    data =b
    c = (
        TreeMap(init_opts=opts.InitOpts(theme=ThemeType.LIGHT, width='800px', height='400px'))
        .add("商品类目", data, )
        .set_global_opts(title_opts=opts.TitleOpts(title="{}-top20商品类目".format(be)))
    )
    return c.render_notebook()

In [271]:
tree_plot('pv')
tree_plot('fav')
tree_plot('cart')
tree_plot('buy')

D:\Anaconda3\lib\site-packages\pyecharts\charts\chart.py:14: PendingDeprecationWarning: pyecharts 所有图表类型将在 v1.9.0 版本开始强制使用 ChartItem 进行数据项配置 :)
  super().__init__(init_opts=init_opts)


D:\Anaconda3\lib\site-packages\pyecharts\charts\chart.py:14: PendingDeprecationWarning: pyecharts 所有图表类型将在 v1.9.0 版本开始强制使用 ChartItem 进行数据项配置 :)
  super().__init__(init_opts=init_opts)


D:\Anaconda3\lib\site-packages\pyecharts\charts\chart.py:14: PendingDeprecationWarning: pyecharts 所有图表类型将在 v1.9.0 版本开始强制使用 ChartItem 进行数据项配置 :)
  super().__init__(init_opts=init_opts)


D:\Anaconda3\lib\site-packages\pyecharts\charts\chart.py:14: PendingDeprecationWarning: pyecharts 所有图表类型将在 v1.9.0 版本开始强制使用 ChartItem 进行数据项配置 :)
  super().__init__(init_opts=init_opts)


通过比较可以发现， pv前20里面也有在其他行为中出现

### 比较所有行为的top20的商品类别

In [280]:
a = [df[df.行为类型==i]['商品类目ID'].value_counts()[0:20] for i in df.行为类型.unique()]

In [282]:
df.行为类型.unique()

array(['pv', 'buy', 'fav', 'cart'], dtype=object)

In [283]:
a[0].index

Int64Index([4756105, 4145813, 2355072, 3607361,  982926, 2520377, 4801426,
            1320293, 2465336, 3002561, 2735466, 4181361,  149192, 1080785,
            2885642, 4217906,  154040, 3738615, 1879194, 2640118],
           dtype='int64')

In [325]:
pd.Series(np.array([a[i].index.tolist() for i in range(4)]).flatten()).value_counts()

2465336    4
4756105    4
2885642    4
2735466    4
1320293    4
982926     4
4801426    4
3002561    4
4145813    4
4217906    4
3607361    4
2520377    3
1879194    3
1464116    3
2355072    3
2640118    3
149192     3
4357323    3
570735     3
4181361    3
1080785    2
3738615    2
154040     1
1299190    1
4789432    1
4159072    1
903809     1
dtype: int64

通过以上对比我们可以发现，pv-top20中前5名的商品在buy-top20中排名分别为5、3、无排名、20、7名，pv-top20、fav-top20、cart-top20的商品品类分别在buy-top20中出现了13、15、15类，说明在点击pv、收藏fav、加购cart环节吸引用户更多注意力的商品没有很好的转化为实际销量，这是销量增长的一个突破点，此外我们也可以看出相比于pv，加购cart、收藏fav与购买的关系更为直接。

# 结论

**一**、通过对用户流量行为的转化分析，可以看出用户从点击到购买的转化率还是比较高的，目前来看可以通过引导用户收藏并加购来提高用户从收藏/加购到购买的转化率。而流量行为从点击到购买的转化率仅有2.24%，故从点击到购买的行为转化是一个提高的重点。针对这一环节的建议有：  
- 1 优化电商平台的搜索匹配度和推荐策略，提高筛选精确度，并对搜索和筛选的结果排序的优先级进行优化；  
- 2 可以给客户提供同类产品比较的功能，让用户不需要多次返回搜索结果进行点击查看，方便用户确定心仪产品，增加点击到后续行为的转化；  
- 3 优化收藏到购买的操作过程，增加用户收藏并加购的频率，以提高购买转化率。  

**二**、通过用户行为习惯分析
- 1 可以看出用户的活跃时间高峰期主要在13:00-14:00点，此时使用人数最多，活动最容易触达用户，所以可以将营销活动安排在这个时间段内，来进行引流并转化； 
  
**三**、通过对用户的商品偏好分析，可以看出商品销量主要是依靠长尾效应，而非爆款商品的带动。但是通过对商品品类的分析可以看出能吸引用户注意力的商品购买转化率并不高，是一个提高销量的突破口。针对用户关注度高但销量不高的这部分产品，可以从以下几个方面着手：
- 1 商家在详情页上的展示突出用户重点关注的信息，优化信息的呈现方式，减少用户的时间成本；  
- 2 增加这些产品的质量管控力度，加强对用户反馈建议如评论的管理，认真采纳并根据自身的优劣势进行商品优化  
- 3 此外，对于购买top20的商品，可以在电商首页对这些品类的商品优先进行展现，以满足用户的购买需求。
